### PR TITLE
Add a check for bundle vendors 

### PR DESCRIPTION
**Pull request by martinvw**
Add a check for proper bundle vendor, I am unsure whether the openHAB repo should contain bundles with eclipse smarthome but it does at this moment (org.openhab.binding.urtsi). 

If this test is NOT useful please let me know as well, thanks!

Signed-off-by: Martin van Wingerden <martinvw@mtin.nl>

### CODE REVIEW

**Comment by martinvw** [comment]
I did not yet link it to the configuration but lets first discuss whether it is indeed useful :-)

File: src/main/java/org/openhab/tools/analysis/checkstyle/ExportInternalPackageCheck.java [hunk 1]
```diff
@@ -24,29 +24,29 @@
  */
 public class ExportInternalPackageCheck extends AbstractStaticCheck {
 
-    public final static String MANIFEST_EXTENSTION = "MF";
+    public final static String MANIFEST_EXTENSION = "MF";
```
Comment (dstivanov) [inline src/main/java/org/openhab/tools/analysis/checkstyle/ExportInternalPackageCheck.java]: I guess that you have looked into this class to write your new check - and thus you have applied some formatting on it. Can you please  revert it back to make the PR clear.

File: src/main/java/org/openhab/tools/analysis/checkstyle/ExportInternalPackageCheck.java [hunk 2]
```diff
@@ -24,29 +24,29 @@
...
     public ExportInternalPackageCheck() {
-        setFileExtensions(MANIFEST_EXTENSTION);
+        setFileExtensions(MANIFEST_EXTENSION);
     }
 
     @Override
     protected void processFiltered(File file, List<String> lines) throws CheckstyleException {
         if (isEmpty(file)) {
-            log(0, MESSAGE_FILE_EMPTY, new Integer(0));
+            log(0, MESSAGE_FILE_EMPTY, 0);
```
Comment (dstivanov) [inline src/main/java/org/openhab/tools/analysis/checkstyle/ExportInternalPackageCheck.java]: I wonder why this was Integer.. Probably @svilenvul can explain it as an author of the check.

File: src/main/java/org/openhab/tools/analysis/checkstyle/BundleVendorCheck.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
+/**
+ * Copyright (c) 2014-2016 by the respective copyright holders.
```
Comment (petarnv) [inline src/main/java/org/openhab/tools/analysis/checkstyle/BundleVendorCheck.java]: I think that the license headers in openHab are now updated to 2017: https://github.com/openhab/openhab2-addons/pull/1865. Can you please change it, so that we follow the same pattern?

File: src/test/java/org/openhab/tools/analysis/checkstyle/test/BundleVendorCheckTest.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+/**
+ * Tests for {@link BundleVendorCheck}
+ *
+ * @author Martin van Wingerden
+ */
+public class BundleVendorCheckTest extends AbstractStaticCheckTest {
+    private static final String TEST_DIRECTORY = "correctOpenHABSpellingInManifestTest/";
+    private DefaultConfiguration checkConfig;
+
+    @Before
```
Comment (petarnv) [inline src/test/java/org/openhab/tools/analysis/checkstyle/test/BundleVendorCheckTest.java]: Don't you use the same configuration for every test? If that is the case can you please make the annotation of the method `@BeforeClass`, so that we create this configuration just one time?

File: src/main/java/org/openhab/tools/analysis/checkstyle/ExportInternalPackageCheck.java (hunk 2 above)
Comment (svilenvul) [inline src/main/java/org/openhab/tools/analysis/checkstyle/ExportInternalPackageCheck.java]: This is one of the first checks that we wrote and we wrote it against the Google checkstyle API (which is slightly different and I think it accepted only objects).

File: src/test/java/org/openhab/tools/analysis/checkstyle/test/BundleVendorCheckTest.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
+package org.openhab.tools.analysis.checkstyle.test;
```
Comment (svilenvul) [inline src/test/java/org/openhab/tools/analysis/checkstyle/test/BundleVendorCheckTest.java]: Please add a header

File: src/test/java/org/openhab/tools/analysis/checkstyle/test/BundleVendorCheckTest.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        );
+    }
+
+    private void verify(String fileName, String[] expectedMessages) throws Exception {
+        verify(checkConfig,
+                getPath(TEST_DIRECTORY + fileName),
+                expectedMessages
+        );
+    }
+}
```
Comment (svilenvul) [inline src/test/java/org/openhab/tools/analysis/checkstyle/test/BundleVendorCheckTest.java]: Please add a new line

**Comment by svilenvul** [comment]
@kaikreuzer we have a PR for check that validates the Bundle-Vendor manifest header.

Maybe I miss something, but I can't find any information in the guidelines. Should/Must the Bundle-Vendor  be "openHAB", or any third party organizations could add their name? Could we include this check?

File: src/main/java/org/openhab/tools/analysis/checkstyle/ExportInternalPackageCheck.java (hunk 1 above)
Comment (martinvw) [inline src/main/java/org/openhab/tools/analysis/checkstyle/ExportInternalPackageCheck.java]: I corrected spelling errors, do you want a separate PR for it?

File: src/main/java/org/openhab/tools/analysis/checkstyle/ExportInternalPackageCheck.java (hunk 2 above)
Comment (martinvw) [inline src/main/java/org/openhab/tools/analysis/checkstyle/ExportInternalPackageCheck.java]: Since Java 5 autoboxing exists, so it is unlikely that it was required...

File: src/main/java/org/openhab/tools/analysis/checkstyle/BundleVendorCheck.java (hunk 3 above)
Comment (martinvw) [inline src/main/java/org/openhab/tools/analysis/checkstyle/BundleVendorCheck.java]: Yes of course, I copied my class, we need a PR for all other classes as well :-)

File: src/test/java/org/openhab/tools/analysis/checkstyle/test/BundleVendorCheckTest.java (hunk 4 above)
Comment (martinvw) [inline src/test/java/org/openhab/tools/analysis/checkstyle/test/BundleVendorCheckTest.java]: The test which I saw in the existing code constructed it in each test so I used the same pattern but using an @Before. IMHO it's not that important if it's @Before or @BeforeClass so I can change it.

File: src/main/java/org/openhab/tools/analysis/checkstyle/BundleVendorCheck.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        } else if (bundleVendors.size() > 1) {
+            tooMany = true;
+        }
+
+        int lineNumber = 0;
+        for (String bundleVendor : bundleVendors) {
+            lineNumber = findLineNumber(lines, bundleVendor, lineNumber);
+
+            if (tooMany) {
+                log(lineNumber, "Only one \"Bundle-Vendor\" was expected.");
```
Comment (martinvw) [inline src/main/java/org/openhab/tools/analysis/checkstyle/BundleVendorCheck.java]: It requires way more String objects (because you have to compose them) and imho it will not become more readable...

Example `"Only one \"Bundle-Vendor\" was expected."`

would become:  
``` 
"Only one \"" +BUNDLE_VENDOR + "\" was expected.";
```
 Would lead to 3 additional String objects (although the compiler might resolve this for) while gaining nothing in readability.

File: src/main/java/org/openhab/tools/analysis/checkstyle/BundleVendorCheck.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+/**
+ * Checks if a manifest file contains the expected bundle vendor
+ *
+ * @author Martin van Wingerden
+ */
+public class BundleVendorCheck extends AbstractStaticCheck {
+    private final static String MANIFEST_EXTENSION = "MF";
+    private static final String REQUIRED_PREFIX = "Bundle-Vendor: ";
+    private static List<String> ALLOWED_VALUES = Arrays.asList("Eclipse.org/SmartHome", "openHAB");
```
Comment (martinvw) [inline src/main/java/org/openhab/tools/analysis/checkstyle/BundleVendorCheck.java]: Great, do you already have a class with an example for that?

File: src/main/java/org/openhab/tools/analysis/checkstyle/BundleVendorCheck.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+        int lineNumber = 0;
+        for (String bundleVendor : bundleVendors) {
+            lineNumber = findLineNumber(lines, bundleVendor, lineNumber);
+
+            if (tooMany) {
+                log(lineNumber, "Only one \"Bundle-Vendor\" was expected.");
+            }
+            if (!bundleVendor.startsWith(REQUIRED_PREFIX)) {
+                log(lineNumber, "Expect eg. \"Bundle-Vendor: openHAB\" got \"" + bundleVendor + "\"");
```
Comment (martinvw) [inline src/main/java/org/openhab/tools/analysis/checkstyle/BundleVendorCheck.java]: This specific error is about incorrect capitalization of 'Bundle-Vendor', missing colons, spaces things like that. So it looks like Bundle-Vendor but is not starting with "Bundle-Vendor: "

Comment (martinvw) [inline src/main/java/org/openhab/tools/analysis/checkstyle/BundleVendorCheck.java]: Reusing the strings to the test does imho make the tests less readable and if you somehow accidently empty or otherwise mutilate your string your test remain happy. 

If we really want that can we then please put the tests into the same package so those strings do not have to be public.

File: src/test/java/org/openhab/tools/analysis/checkstyle/test/BundleVendorCheckTest.java (hunk 5 above)
Comment (martinvw) [inline src/test/java/org/openhab/tools/analysis/checkstyle/test/BundleVendorCheckTest.java]: Will do, testing this is the goal of this project :-S

File: src/test/java/org/openhab/tools/analysis/checkstyle/test/BundleVendorCheckTest.java (hunk 6 above)
Comment (martinvw) [inline src/test/java/org/openhab/tools/analysis/checkstyle/test/BundleVendorCheckTest.java]: Will do, testing this is the goal of this project :-S

**Comment by martinvw** [comment]
Thanks for the feedback I have some replies on things please let me know what is important for you to accept it. I will definitely fix the style errors.

File: src/main/java/org/openhab/tools/analysis/checkstyle/BundleVendorCheck.java (hunk 3 above)
Comment (petarnv) [inline src/main/java/org/openhab/tools/analysis/checkstyle/BundleVendorCheck.java]: Yes, some of the other classes need to be updated, too. :) 
Thanks for noticing.

File: src/test/java/org/openhab/tools/analysis/checkstyle/test/BundleVendorCheckTest.java (hunk 4 above)
Comment (petarnv) [inline src/test/java/org/openhab/tools/analysis/checkstyle/test/BundleVendorCheckTest.java]: Whatever you decide.

File: src/main/java/org/openhab/tools/analysis/checkstyle/BundleVendorCheck.java (hunk 8 above)
Comment (dstivanov) [inline src/main/java/org/openhab/tools/analysis/checkstyle/BundleVendorCheck.java]: Yes! not yet merged, but present in a PR. See here: https://github.com/openhab/static-code-analysis/pull/25/files#diff-82672bd78945c746d7c6386e8df15080R45

File: src/main/java/org/openhab/tools/analysis/checkstyle/BundleVendorCheck.java (hunk 7 above)
Comment (dstivanov) [inline src/main/java/org/openhab/tools/analysis/checkstyle/BundleVendorCheck.java]: Your argument is good, so I agree. leave it like that :)

File: src/main/java/org/openhab/tools/analysis/checkstyle/BundleVendorCheck.java (hunk 9 above)
Comment (dstivanov) [inline src/main/java/org/openhab/tools/analysis/checkstyle/BundleVendorCheck.java]: I got your point, in the way it is now we will have passing tests even if we have incorrect messages. So yes, it is better to leave the tests with hardcoded messages,

File: src/main/java/org/openhab/tools/analysis/checkstyle/ExportInternalPackageCheck.java (hunk 1 above)
Comment (dstivanov) [inline src/main/java/org/openhab/tools/analysis/checkstyle/ExportInternalPackageCheck.java]: No, I don't want separate PR, I just missed the spelling error. Thanks for fixing them!

**Comment by kaikreuzer** [comment]
> Should/Must the Bundle-Vendor be "openHAB", or any third party organizations could add their name? Could we include this check?

Yes, this check makes a lot of sense. Our archetype creates it as "openHAB" and it must not be changed for any official openHAB bundles.

**Comment by martinvw** [comment]
The is also one binding in the openhab project which has eclipse smarthome as bundle vendor, should this be fixed? See text of the pr.

Should i also include (some) other default from the archetype, if yes, which?

I'll work on the comments later so that is ready to be merged.

**Comment by kaikreuzer** [comment]
> The is also one binding in the openhab project which has eclipse smarthome as bundle vendor, should this be fixed?

Yes, please.

> Should i also include (some) other default from the archetype, if yes, which?

I don't see anything specific to add.

**Comment by martinvw** [comment]
All done:

- added the missing file header
- updated the year in the copyright header 
- added a new line at the end of the file
- made the allowed values configurable 
- use a BeforeClass instead of normal Before.
- created a PR for the OH2 binding with the wrong vendor

IMHO all should be fine now.

**Comment by martinvw** [comment]
@kaikreuzer should the check for Java 8 also be included in this check it seems like a logical place for it? 

The class could then use some renaming but that shouldn't be a problem.

**Comment by dstivanov** [comment]
@martinvw We are having several more checks for the Manifest as well, so I guess that the Java version can be added into separate check. By having the checks more distributed we will have the freedom to control whether they are executed or not, without having to alter the code, but only their configurations.

**Comment by svilenvul** [comment]
Thanks !
